### DISCUSSION TITLE
Does LightningCLI support user-defined subcommands?

### DISCUSSION THREAD

**Discussion by cgoliver** [Lightning-AI/pytorch-lightning#14314] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I would like to define some new subcommands along with the default ones and be able to attach some function calls to it. The documentation does not seem to show this use case so I'm wondering if this is possible.

For example, to visualize the output of some model.

```
python main.py visualize --checkpoint=/my/checkpoing.ckpt
```

Or would it be best to just customize the behaviour of some existing subcommands as described [here](https://pytorch-lightning.readthedocs.io/en/stable/cli/lightning_cli_expert.html#customize-the-lightningcli)?

**Comment by mauvilsa** [comment, marked as answer]
To define new subcommands I would recommend the following. First you subclass `Trainer` and add methods for each of the subcommands you want. These new methods should be usable independent of the CLI. Then subclass `LightningCLI`, in the `__init__` make your new trainer the default and to add the subcommands simply override/extend https://github.com/Lightning-AI/lightning/blob/7a617ec90e1566c763be8ac7a200af1e4025412c/src/pytorch_lightning/cli.py#L425-L434

**Reply by mauvilsa** [reply]
> The user wants to have the output of the model (a batch_size by dim embeddings) clustered in order to use them.

I did not understand this.

**Reply by cgoliver** [reply]
That was just intended as an example of a subcommand that would do more than just output the raw model predictions. It was confusing and have removed it from the question. Thanks for the answer!
